Release ouroboros-consensus-diffusion 0.4

Jasagredo opened the pull request.

amesgen commented on ouroboros-consensus-byron/ouroboros-consensus-byron.cabal:
```diff
@@ -90,7 +90,7 @@ executable db-converter
     , ouroboros-consensus            ^>=0.4
     , ouroboros-consensus-byron
-    , ouroboros-consensus-diffusion  ^>=0.4
+    , ouroboros-consensus-diffusion  (^>=0.4 || ^>=0.5)
```
I was going to suggest
```haskell
    , ouroboros-consensus-diffusion  ^>=0.4 || ^>=0.5
```
but `cabal-fmt` turns that into
```haskell
    , ouroboros-consensus-diffusion  >=0.4  && <0.5  || ^>=0.5
```
right?

coot replied:
why not just:
```
, ouroboros-consensus-diffusion  >=0.4 && <0.6
```

amesgen reviewed (changes requested):
There are still a few constraints that force `ouroboros-consensus-diffusion == 0.4.*`.

Jasagredo commented:
I know there are references here and there. To make sure it built I removed the network packages and then built the main implementation line without the tests.
```
modified   cabal.project
@@ -18,34 +18,15 @@ index-state:
   -- Bump this if you need newer packages from Hackage
   , hackage.haskell.org 2023-03-29T00:00:00Z
   -- Bump this if you need newer packages from CHaP
-  , cardano-haskell-packages 2023-04-05T21:25:44Z
+  , cardano-haskell-packages 2023-04-25T00:00:00Z
 
-packages: ./cardano-ping
-          ./monoidal-synchronisation
-          ./network-mux
-          ./ouroboros-network
-          ./ouroboros-network-api
-          ./ouroboros-network-framework
-          ./ouroboros-network-mock
-          ./ouroboros-network-protocols
-          ./ouroboros-network-testing
-          ./ouroboros-consensus
-          ./ouroboros-consensus-tutorials
+packages: ./ouroboros-consensus
           ./ouroboros-consensus-byron
-          ./ouroboros-consensus-byron-test
-          ./ouroboros-consensus-byronspec
           ./ouroboros-consensus-cardano
-          ./ouroboros-consensus-cardano-test
-          ./ouroboros-consensus-mock
-          ./ouroboros-consensus-mock-test
           ./ouroboros-consensus-protocol
           ./ouroboros-consensus-shelley
-          ./ouroboros-consensus-shelley-test
-          ./ouroboros-consensus-test
           ./ouroboros-consensus-cardano-tools
           ./ouroboros-consensus-diffusion
-          ./ntp-client
-          ./cardano-client
```

But probably CI doesn't allow me to do this, so I will fix the others even if we don't release them now

coot commented on ouroboros-consensus-diffusion/CHANGELOG.md:
```diff
@@ -1,5 +1,43 @@
+  - Monomorphized `Handlers` client field types to `ConnectionId addr`;
+    and added PeerSharing handlers.
+  - Changed `mkHandlers` function to receive a function to compute peer sharing addresses;
+  - Changed `Codecs` type and propagated changes to relevant functions (e.g. `defaultCodecs`);
```
@bolt12 haven't we renamed `defaultCodecs` to `serialiseCodecs`?

bolt12 replied:
No that PR hasn't been merged yet, I forgot to address that bit

coot replied:
Ignore this comment, your `defaultCodecs` is something different.

coot reviewed (commented):
LGTM, just one minor comment.

Jasagredo commented:
bors r+
